### PR TITLE
bluetooth: mesh: ensure input_complete callback in provisioner

### PR DESCRIPTION
**Pull request by akredalen**
Call input_complete whenever the link is invalidated or provisioning succeeds. The application may be awaiting this callback.

Signed-off-by: Stine Akredalen <stine.akredalen@nordicsemi.no>

### CODE REVIEW

**Review by alxelax — CHANGES_REQUESTED** [review]
It doesn't have sense for provisioning without OOB or with OOB with different to OOB input. 
That causes NULL dereferencing. 
I'm not sure in that BLE Mesh has to call this callback at all. Seems if application gets stuck forever in a waiting of input complete this is bad design of application. BLE Mesh notifies about link closing with statuses and application should make decision about provisioning result basing on that.

**Comment by PavelVPV** [comment]
> It doesn't have sense for provisioning without OOB or with OOB with different to OOB input. That causes NULL dereferencing. I'm not sure in that BLE Mesh has to call this callback at all. Seems if application gets stuck forever in a waiting of input complete this is bad design of application. BLE Mesh notifies about link closing with statuses and application should make decision about provisioning result basing on that.

I tend to agree. An application may have some additional functionality triggered by `input_complete` callback, but this will have to be aborted anyway if provisioning failed. We have `link_close` callback which must always be called by the end of provisioning. This callback can be used to abort any behavior started in application during the provisioning.

**Comment by ludvigsj** [comment]
> > It doesn't have sense for provisioning without OOB or with OOB with different to OOB input. That causes NULL dereferencing. I'm not sure in that BLE Mesh has to call this callback at all. Seems if application gets stuck forever in a waiting of input complete this is bad design of application. BLE Mesh notifies about link closing with statuses and application should make decision about provisioning result basing on that.
> 
> I tend to agree. An application may have some additional functionality triggered by `input_complete` callback, but this will have to be aborted anyway if provisioning failed. We have `link_close` callback which must always be called by the end of provisioning. This callback can be used to abort any behavior started in application during the provisioning.

I think the idea was that the `input_complete`-callback should be the only place the application needs to perform the "stop displaying output value" behavior. This is what the callback is described as in the documentation. (I.e., the API docs present `input_complete` as a callback meaning exclusively "Application, please stop displaying output value", which you would also need to do upon abort).

I see your point though, the only sensible thing to do on the application side if provisioning fails is to reset all state, including any displayed output value, even if the API docs could be read like you may depend on input complete.

**Comment by alxelax** [comment]
> > > It doesn't have sense for provisioning without OOB or with OOB with different to OOB input. That causes NULL dereferencing. I'm not sure in that BLE Mesh has to call this callback at all. Seems if application gets stuck forever in a waiting of input complete this is bad design of application. BLE Mesh notifies about link closing with statuses and application should make decision about provisioning result basing on that.
> > 
> > 
> > I tend to agree. An application may have some additional functionality triggered by `input_complete` callback, but this will have to be aborted anyway if provisioning failed. We have `link_close` callback which must always be called by the end of provisioning. This callback can be used to abort any behavior started in application during the provisioning.
> 
> I think the idea was that the `input_complete`-callback should be the only place the application needs to perform the "stop displaying output value" behavior. This is what the callback is described as in the documentation. (I.e., the API docs present `input_complete` as a callback meaning exclusively "Application, please stop displaying output value", which you would also need to do upon abort).
> 
> I see your point though, the only sensible thing to do on the application side if provisioning fails is to reset all state, including any displayed output value, even if the API docs could be read like you may depend on input complete.

I would leave it up to application developer responsibility. I guess Mesh shouldn't try to predict application behavior.

**Comment by PavelVPV** [comment]
> > > > It doesn't have sense for provisioning without OOB or with OOB with different to OOB input. That causes NULL dereferencing. I'm not sure in that BLE Mesh has to call this callback at all. Seems if application gets stuck forever in a waiting of input complete this is bad design of application. BLE Mesh notifies about link closing with statuses and application should make decision about provisioning result basing on that.
> > > 
> > > 
> > > I tend to agree. An application may have some additional functionality triggered by `input_complete` callback, but this will have to be aborted anyway if provisioning failed. We have `link_close` callback which must always be called by the end of provisioning. This callback can be used to abort any behavior started in application during the provisioning.
> > 
> > 
> > I think the idea was that the `input_complete`-callback should be the only place the application needs to perform the "stop displaying output value" behavior. This is what the callback is described as in the documentation. (I.e., the API docs present `input_complete` as a callback meaning exclusively "Application, please stop displaying output value", which you would also need to do upon abort).
> > I see your point though, the only sensible thing to do on the application side if provisioning fails is to reset all state, including any displayed output value, even if the API docs could be read like you may depend on input complete.
> 
> I would leave it up to application developer responsibility. I guess Mesh shouldn't try to predict application behavior.

Agree. `input_complete` is called when the device _finishes_ OOB input:
https://github.com/zephyrproject-rtos/zephyr/blob/b8a957a0ebd955ed65b7f1e64c74dac794a37fb3/include/zephyr/bluetooth/mesh/main.h#L194
If the input wasn't finished (there were no `Input Complete` PDU from the device), the callback should not be called.

**Comment by akredalen** [comment]
The documentation reads: "When an Output OOB action is selected, the authentication value should be presented to the user when the output callback is called, and remain until the :c:member:`bt_mesh_prov.input_complete` or :c:member:`bt_mesh_prov.complete` callback is called." This implies that the application can rely on either the `input_complete` or `complete` callback to be called, however, this is not the case if the provisioning link is closed due to some failure.

**Bottom line**: I agree that mesh shouldn't try to predict application behavior and that the link_close callback can be used to abort any behavior started in the application during the provisioning.  I do suggest however that the documentation is updated to avoid future issues with this in applications.

**Comment by PavelVPV** [comment]
> The documentation reads: "When an Output OOB action is selected, the authentication value should be presented to the user when the output callback is called, and remain until the :c:member:`bt_mesh_prov.input_complete` or :c:member:`bt_mesh_prov.complete` callback is called." This implies that the application can rely on either the `input_complete` or `complete` callback to be called, however, this is not the case if the provisioning link is closed due to some failure.
> 
> **Bottom line**: I agree that mesh shouldn't try to predict application behavior and that the `link_close` callback can be used to abort any behavior started in the application during the provisioning.  I do suggest however that the documentation is updated to avoid future issues with this in applications.

I guess that the paragraph describes a successful case. But I agree, the provisioning documentation could be more descriptive and describe the case when the provisioning fails.
